### PR TITLE
[cdc-connector][mysql]Mysql add numRecordsOutBySnapshot and numRecordsOutByIncremental  metrics

### PR DESCRIPTION
**Pull request by ChengJie1053**
`numRecordsOutBySnapshot` metrics collects snapshot statistics

`numRecordsOutByIncremental` metrics collects incremental statistics

`numRecordsOutByIncremental` has three metrics, respectively is `numRecordsOutByIncrementalInsert`, `numRecordsOutByIncrementalUpdate`, `numRecordsOutByIncrementalDelete`

image

2024.7.12 change metrics name，change `numRecordsOutByIncrementalInsert` to  `numRecordsOutByDataChangeEventInsert`
image

### CODE REVIEW

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/reader/MySqlRecordEmitter.java [hunk 1]
```diff
@@ -120,7 +123,31 @@ private void emitElement(SourceRecord element, SourceOutput<T> output) throws Ex
...
+                            break;
+                        case UPDATE:
+                            sourceReaderMetrics.numRecordsOutUpdateIncrease(tableId);
+                            break;
+                        case DELETE:
+                            sourceReaderMetrics.numRecordsOutDeleteIncrease(tableId);
+                            break;
+                    }
+                }
+            }
```
Comment (SML0127) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/reader/MySqlRecordEmitter.java:149]: We checked whether element is data change record or not in line 101.
How about removing this condition?

```suggestion
            Struct value = (Struct) element.value();
            if (value != null) {
                Envelope.Operation operation =
                        Envelope.Operation.forCode(
                                value.getString(Envelope.FieldName.OPERATION));
                switch (operation) {
                    case CREATE:
                        sourceReaderMetrics.numRecordsOutInsertIncrease(tableId);
                        break;
                    case UPDATE:
                        sourceReaderMetrics.numRecordsOutUpdateIncrease(tableId);
                        break;
                    case DELETE:
                        sourceReaderMetrics.numRecordsOutDeleteIncrease(tableId);
                        break;
                }
            }
```

Comment (ChengJie1053) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/reader/MySqlRecordEmitter.java:149]: Thank you very much for helping me review the code.
I will delete the isDataChangeRecord judgment

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/metrics/MySqlSourceReaderMetrics.java [hunk 2]
```diff
@@ -17,18 +17,37 @@
...
+    public static final String IO_NUM_RECORDS_OUT_SNAPSHOT = ".numRecordsOutBySnapshot";
+
+    public static final String IO_NUM_RECORDS_OUT_INCREMENTAL_INSERT =
+            ".numRecordsOutByIncrementalInsert";
+
+    public static final String IO_NUM_RECORDS_OUT_INCREMENTAL_DELETE =
+            ".numRecordsOutByIncrementalDelete";
+
+    public static final String IO_NUM_RECORDS_OUT_INCREMENTAL_UPDATE =
+            ".numRecordsOutByIncrementalUpdate";
```
[the remark this answers is not in the document]
Comment (ChengJie1053) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/metrics/MySqlSourceReaderMetrics.java:47]: I think your suggestion is great, and I will CHANGE it to CHANGE EVENT

**Review by SML0127 — APPROVED** [review]
LGTM!

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/metrics/MySqlSourceReaderMetrics.java [hunk 3]
```diff
@@ -44,6 +69,49 @@ public void registerMetrics() {
...
+        Counter counter =
+                deleteNumRecordsOutMap.computeIfAbsent(
+                        tableId,
+                        k ->
+                                metricGroup.counter(
+                                        tableId.identifier()
+                                                + IO_NUM_RECORDS_OUT_DATA_CHANGE_EVENT_DELETE));
+        counter.inc();
+    }
+
```
Comment (SML0127) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/metrics/MySqlSourceReaderMetrics.java:114]: I applied this metrics my code, and left some comments.

- Metric names should be general. 
    - In your case, It generate uniqe metric for each tables. But in monitorig tool such as grafana, we distingush tables by property in metric, not by mtric name
So I suggest to remove tableId.   
```suggestion
    public void numRecordsOutSnapshotIncrease(TableId tableId) {
        Counter counter =
                snapshotNumRecordsOutMap.computeIfAbsent(
                        tableId,
                        k ->
                                metricGroup.counter(IO_NUM_RECORDS_OUT_SNAPSHOT));
        counter.inc();
    }

    public void numRecordsOutInsertIncrease(TableId tableId) {
        Counter counter =
                insertNumRecordsOutMap.computeIfAbsent(
                        tableId,
                        k ->
                                metricGroup.counter(IO_NUM_RECORDS_OUT_DATA_CHANGE_EVENT_INSERT));
        counter.inc();
    }

    public void numRecordsOutUpdateIncrease(TableId tableId) {
        Counter counter =
                updateNumRecordsOutMap.computeIfAbsent(
                        tableId,
                        k ->
                                metricGroup.counter(IO_NUM_RECORDS_OUT_DATA_CHANGE_EVENT_UPDATE));
        counter.inc();
    }

    public void numRecordsOutDeleteIncrease(TableId tableId) {
        Counter counter =
                deleteNumRecordsOutMap.computeIfAbsent(
                        tableId,
                        k ->
                                metricGroup.counter(IO_NUM_RECORDS_OUT_DATA_CHANGE_EVENT_DELETE));
        counter.inc();
    }
```

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/metrics/MySqlSourceReaderMetrics.java [hunk 4]
```diff
@@ -18,17 +18,42 @@
...
+    public static final String IO_NUM_RECORDS_OUT_SNAPSHOT = ".numRecordsOutBySnapshot";
+
+    public static final String IO_NUM_RECORDS_OUT_DATA_CHANGE_EVENT_INSERT =
+            ".numRecordsOutByDataChangeEventInsert";
+
+    public static final String IO_NUM_RECORDS_OUT_DATA_CHANGE_EVENT_DELETE =
+            ".numRecordsOutByDataChangeEventDelete";
+
+    public static final String IO_NUM_RECORDS_OUT_DATA_CHANGE_EVENT_UPDATE =
+            ".numRecordsOutByDataChangeEventUpdate";
```
Comment (SML0127) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/metrics/MySqlSourceReaderMetrics.java:47]: I suggest to remove dot(.) from metric name.

```suggestion
    public static final String IO_NUM_RECORDS_OUT_SNAPSHOT = "numRecordsOutBySnapshot";

    public static final String IO_NUM_RECORDS_OUT_DATA_CHANGE_EVENT_INSERT =
            "numRecordsOutByDataChangeEventInsert";

    public static final String IO_NUM_RECORDS_OUT_DATA_CHANGE_EVENT_DELETE =
            "numRecordsOutByDataChangeEventDelete";

    public static final String IO_NUM_RECORDS_OUT_DATA_CHANGE_EVENT_UPDATE =
            "numRecordsOutByDataChangeEventUpdate";
```

**Review by SML0127 — CHANGES_REQUESTED** [review]
PTAL

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/metrics/MySqlSourceReaderMetrics.java (hunk 4 above)
Comment (ChengJie1053) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/metrics/MySqlSourceReaderMetrics.java:47]: Ok, thank you for helping me review the code, I will remove (.)

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/metrics/MySqlSourceReaderMetrics.java (hunk 3 above)
Comment (ChengJie1053) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/metrics/MySqlSourceReaderMetrics.java:114]: Ok, thanks for reviewing the code for me, but the original purpose of this metric is to see each table at a granular level, if you need to see the total, I will add other metrics

**Comment by yuxiqian** [comment]
Seems we can't come up with a "best" metric design for everyone: Some may want a task-level unified record counter, while some may want fine-grained counters for each table. Or one may want to disable it entirely to avoid any extra cost.

I wonder if we should add an option to control metric behavior?

**Comment by SML0127** [comment]
@ChengJie1053 cc. @yuxiqian 
Is thera any changes or progress? If not, I want to create new PR with my suggestion at next week. 
(I need this feature.)

**Pull request closed without merging**